src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs

```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
 
             stateMachineType = new IteratorStateMachine(slotAllocatorOpt, compilationState, method, methodOrdinal, isEnumerable, elementType);
             compilationState.ModuleBuilderOpt.CompilationState.SetStateMachineType(method, stateMachineType);
-            return new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics).Rewrite();
+            var rewriter = new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics);
+            if (!rewriter.VerifyPresenceOfRequiredAPIs())
+            {
+                return body;
+            }
+
+            return rewriter.Rewrite();
+        }
+
+        /// <returns>
+        /// Returns true if all types and members we need are present and good
+        /// </returns>
+        protected bool VerifyPresenceOfRequiredAPIs()
+        {
+            DiagnosticBag bag = DiagnosticBag.GetInstance();
+            try
+            {
+                EnsureSpecialType(SpecialType.System_Int32, bag);
+                EnsureSpecialType(SpecialType.System_IDisposable, bag);
+                EnsureSpecialMember(SpecialMember.System_IDisposable__Dispose, bag);
+
+                // IEnumerator
+                EnsureSpecialType(SpecialType.System_Collections_IEnumerator, bag);
+                EnsureSpecialPropertyGetter(SpecialMember.System_Collections_IEnumerator__Current, bag);
+                EnsureSpecialMember(SpecialMember.System_Collections_IEnumerator__MoveNext, bag);
+                EnsureSpecialMember(SpecialMember.System_Collections_IEnumerator__Reset, bag);
+
+                // IEnumerator<T>
+                EnsureSpecialType(SpecialType.System_Collections_Generic_IEnumerator_T, bag);
+                EnsureSpecialPropertyGetter(SpecialMember.System_Collections_Generic_IEnumerator_T__Current, bag);
+
+                if (_isEnumerable)
+                {
+                    // IEnumerable and IEnumerable<T>
+                    EnsureSpecialType(SpecialType.System_Collections_IEnumerable, bag);
+                    EnsureSpecialMember(SpecialMember.System_Collections_IEnumerable__GetEnumerator, bag);
+                    EnsureSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T, bag);
+                    EnsureSpecialMember(SpecialMember.System_Collections_Generic_IEnumerable_T__GetEnumerator, bag);
+                }
+
+                if (bag.HasAnyErrors())
+                {
+                    diagnostics.AddRange(bag);
+                    return false;
+                }
+
+                return true;
+            }
+            finally
+            {
+                bag.Free();
+            }
```

Comment:
```try ... finally``` is typically avoided for freeing pooled instances. Instead consider something like the following here and in VB.
```
bool hasErrors = bag.HasAnyErrors();
if (hasErrors)
{
    diagnostics.AddRange(bag);
}
bag.Free();
return !hasErrors;
```